temporal_sqlalchemy/bases.py

```diff
@@ -9,18 +9,71 @@
 import sqlalchemy as sa
 import sqlalchemy.dialects.postgresql as sap
 import sqlalchemy.orm as orm
+import sqlalchemy.orm.base as base
 import sqlalchemy.orm.attributes as attributes
 import psycopg2.extras as psql_extras
 
 from temporal_sqlalchemy import nine
-from temporal_sqlalchemy.metadata import get_session_metadata
 
-_ClockSet = collections.namedtuple('_ClockSet', ('effective', 'vclock'))
+_PersistentClockPair = collections.namedtuple('_PersistentClockPairs',
+                                              ('effective', 'vclock'))
 
 T_PROPS = typing.TypeVar(
     'T_PROP', orm.RelationshipProperty, orm.ColumnProperty)
 
 
+class ActivityState:
+    def __set__(self, instance, value):
+        assert instance.temporal_options.activity_cls, "Make this better Joey"
+        # TODO should not be able to change activity once changes have been made to temporal properties
+        setattr(instance, '__temporal_current_activity', value)
+
+        if value:
+            current_clock = instance.current_clock
+            current_clock.activity = value
+
+    def __get__(self, instance, owner):
+        if not instance:
+            return self
+
+        return getattr(instance, '__temporal_current_activity', None)
+
+    @staticmethod
+    def reset_activity(target, attr):
+        target.activity = None
+
+    @staticmethod
+    def activity_required(target, value, oldvalue, initiator):
+        if not target.activity and oldvalue is not base.NEVER_SET:
+            raise ValueError("activity required")
+
+
+class ClockState:
+    def __set__(self, instance, value: 'EntityClock'):
+        setattr(instance, '__temporal_current_tick', value)
+        if value:
+            instance.clock.append(value)
+
+    def __get__(self, instance, owner):
+        if not instance:
+            return self
+
+        vclock = getattr(instance, 'vclock') or 0
+        if not getattr(instance, '__temporal_current_tick', None):
+            new_version = vclock + 1
+            instance.vclock = new_version
+            clock_tick = instance.temporal_options.clock_model(tick=new_version)
+            setattr(instance, '__temporal_current_tick', clock_tick)
+            instance.clock.append(clock_tick)
+
+        return getattr(instance, '__temporal_current_tick')
```

Comment:
Same note here about using `getattr` without a default argument.

_However_, be warned: attributes starting with double underscores may not behave as you expect. Python has a concept of "[private variables](https://docs.python.org/3/tutorial/classes.html#tut-private)" that may result in an `AttributeError` getting thrown.

From the docs:

> Any identifier of the form `__spam` (at least two leading underscores, at most one trailing underscore) is textually replaced with `_classname__spam`, where classname is the current class name with leading underscore(s) stripped. This mangling is done without regard to the syntactic position of the identifier, as long as it occurs within the definition of a class.

Because it's required to occur inside the class definition I think this will be okay, but I thought I'd point it out.

Reply:
yeah, I think in this case it's okay, but it's a good call out